Missing type inference for generic trait implementation

```rust
#[lang = "sized"]
trait Sized {}

struct Foo {
    t: u64,
}

impl Foo {
    fn of<T>() -> Foo {
        Foo { t: 14 }
    }
}

trait Bar {
    fn bar() -> Foo;
}

impl<T> Bar for T {
    fn bar() -> Foo {
        Foo::of::<T>()
    }
}
```

`rustc` shows no error for this code, as the implementation is completely generic - but we do raise an error which is invalid:

```rust
arthur@platypus ~/G/r/gccrs (master)> build/gcc/crab1 test.rs
test.rs:20:14: error: type annotations needed [E0282]
   20 |         Foo::of::<T>()
      |              ^~
```

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -490,14 +490,6 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
     }
 
   rust_assert (resolved_node_id != UNKNOWN_NODEID);
-  if (tyseg->needs_generic_substitutions () && !reciever_is_generic)
-    {
-      location_t locus = segments.back ().get_locus ();
-      tyseg = SubstMapper::InferSubst (tyseg, locus);
-      if (tyseg->get_kind () == TyTy::TypeKind::ERROR)
-	return;
-    }
-
   context->insert_receiver (expr_mappings.get_hirid (), prev_segment);
 
   // name scope first

--- a/gcc/testsuite/rust/compile/issue-3009.rs
+++ b/gcc/testsuite/rust/compile/issue-3009.rs
@@ -1,0 +1,24 @@
+#[lang = "sized"]
+trait Sized {}
+
+struct Foo {
+    // { dg-warning "struct is never constructed" "" { target *-*-* } .-1 }
+    t: u64,
+}
+
+impl Foo {
+    fn of<T>() -> Foo {
+        // { dg-warning "associated function is never used" "" { target *-*-* } .-1 }
+        Foo { t: 14 }
+    }
+}
+
+trait Bar {
+    fn bar() -> Foo;
+}
+
+impl<T> Bar for T {
+    fn bar() -> Foo {
+        Foo::of::<T>()
+    }
+}

--- a/gcc/testsuite/rust/compile/nr2/exclude
+++ b/gcc/testsuite/rust/compile/nr2/exclude
@@ -221,4 +221,5 @@ issue-3032-2.rs
 if_let_expr_simple.rs
 iflet.rs
 issue-3033.rs
+issue-3009.rs
 # please don't delete the trailing newline

